tool/ruby_vm/views/_insn_entry.erb

```diff
@@ -33,12 +37,29 @@ INSN_ENTRY(<%= insn.name %>)
 % if insn.handles_sp?
     POPN(INSN_ATTR(popn));
 % end
+#ifdef _FORTIFY_SOURCE
+    if ((leaf = INSN_ATTR(leaf))) {
+        canary = STACK_ADDR_FROM_TOP(-1);
+        *canary = rb_vm_stack_canary;
+    }
+#endif
     COLLECT_USAGE_INSN(INSN_ATTR(bin));
 % insn.opes.each_with_index do |ope, i|
     COLLECT_USAGE_OPERAND(INSN_ATTR(bin), <%= i %>, <%= ope[:name] %>);
 % end
 <%= render_c_expr insn.expr -%>
     CHECK_VM_STACK_OVERFLOW_FOR_INSN(VM_REG_CFP, INSN_ATTR(retn));
+% unless insn.complicated_return_values?
+#ifdef _FORTIFY_SOURCE
```

Comment:
how about to check at `vm_push_frame` with checking the instruction type?
like that:
```
def vm_push_frame(cfp, ...)
  if leaf_insn(cfp->pc) fail
```

Reply:
This is possible.  The reason I avoided checking in vm_push_frame was that

1.  I was not sure if this is _the_ only place stack is smashed -- especially when interrupted.
2. We need a little more code than the proposed patch to implement "if leaf_insn(cfp->pc)" detector (canary is embedded inside of the instruction itself now).

But given we are going to use VM_CHECK_MODE, why not implement both canary and check at vm_push_frame.  That is kind of redundant but not immediately a bad thing here.